args/core/scheduling/scheduler.hpp

```diff
@@ -12,208 +13,272 @@
+    public:
+        struct runnable
+        {
+            runnable() = default;
+            runnable(const delegate<void(void*)>& func, void* param) : func(func), param(param) {}
+
+            delegate<void(void*)> func;
+            void* param;
+
+            void operator()(void)
+            {
+                if (func != nullptr)
+                    func(param);
+            }
+        };
+
+    private:
+        struct thread_error
+        {
+            std::string message;
+            std::thread::id threadId;
+        };
+
+        ProcessChain m_localChain;
+        async::readonly_rw_spinlock m_processChainsLock;
+        sparse_map<id_type, ProcessChain> m_processChains;
+        sparse_map<id_type, std::thread::id> m_chainThreads;
+        async::readonly_rw_spinlock m_errorsLock;
+        std::vector<thread_error> m_errors;
+
+        async::readonly_rw_spinlock m_exitsLock;
+        std::vector<std::thread::id> m_exits;
+
+        std::atomic_bool m_requestSync;
+        async::ring_sync_lock m_syncLock;
+
+        std::atomic<float> m_timeScale = 1.f;
+
+        events::EventBus* m_eventBus;
+
+        bool m_threadsShouldTerminate = false;
+        bool m_threadsShouldStart = false;
+        bool m_low_power;
+
+        static async::readonly_rw_spinlock m_threadsLock;
+        static sparse_map<std::thread::id, std::unique_ptr<std::thread>> m_threads;
+        static std::queue<std::thread::id> m_unreservedThreads;
+        static const uint m_maxThreadCount;
+        static async::readonly_rw_spinlock m_availabilityLock;
+        static uint m_availableThreads;
+
+        static async::readonly_rw_spinlock m_jobQueueLock;
+        static std::queue<runnable> m_jobs;
+        static sparse_map<std::thread::id, async::readonly_rw_spinlock> m_commandLocks;
+        static sparse_map<std::thread::id, std::queue<Scheduler::runnable>> m_commands;
+
+        static void threadMain(bool* exit, bool* start, bool low_power);
+
+    public:
+        Scheduler(events::EventBus* eventBus, bool low_power);
+
+        ~Scheduler();
+
+        /**@brief Set global time scale.
+         */
+        void setTimeScale(float scale)
+        {
+            m_timeScale.store(scale, std::memory_order_relaxed);
+        }
+
+        /**@brief Get the global time scale.
+         */
+        float getTimeScale()
+        {
+            return m_timeScale.load(std::memory_order_relaxed);
+        }
+
+        /**@brief Run main program loop, also starts all process-chains in their own threads.
+         */
+        void run();
+
+        /**@brief Create a new thread.
+         * @param function Function to run on the thread.
+         * @param ...args Arguments to pass to the function.
+         * @return std::thread::id Id of new thread if thread was created, 0 id if there are no available threads.
+         */
+        template<typename Function, typename... Args >
+        std::thread::id createThread(Function&& function, Args&&... args)
```

Comment:
consider making this return `common::result<std::thread::id,void>` or introducing an alias `invalid_thread_id` for better readability of error states `std::thread::id()` does not seem like it indicates graceful-failure